util/merkle-tree/src/merkle_tree.rs

```diff
@@ -0,0 +1,354 @@
+        ((num + 1) ^ 1) - 1
+    }
+}
+
+pub fn calc_parent(num: usize) -> usize {
+    if num == 0 {
+        0
+    } else {
+        (num - 1) >> 1
+    }
+}
+
+pub fn is_left(num: usize) -> bool {
+    num & 1 == 1
+}
+
+#[derive(Eq, PartialEq, Debug, Serialize, Deserialize, Clone)]
+pub struct CombinedProof {
+    // size of items in the tree
+    pub size: u32,
+    // nodes on the path which can not be calculated, in descending order by index
+    pub nodes: Vec<H256>,
+}
+
+impl CombinedProof {
+    pub fn root(&self, mut tuples: Vec<(u32, H256)>) -> Option<H256> {
+        let mut queue = VecDeque::new();
+        tuples.sort_by(|a, b| a.0.cmp(&b.0));
+        tuples.reverse();
+
+        for (index, hash) in tuples {
+            queue.push_back(((index + self.size - 1) as usize, hash));
+        }
+
+        let mut nodes_iter = self.nodes.iter();
+
+        while !queue.is_empty() {
+            let mut in_queue = false;
+
+            let (index1, hash1) = queue.pop_front().unwrap();
+            let sibling = calc_sibling(index1);
+
+            if let Some((index2, _)) = queue.front() {
+                if index2 == &sibling {
+                    in_queue = true;
+                }
+            }
+
+            let hash2 = if in_queue {
+                queue.pop_front().unwrap().1
+            } else if let Some(h) = nodes_iter.next() {
+                h.clone()
+            } else {
+                return None;
+            };
+
+            let hash = if is_left(index1) {
+                merge(&hash1, &hash2)
+            } else {
+                merge(&hash2, &hash1)
+            };
+
+            let parent = calc_parent(index1);
+
+            if parent == 0 {
+                return Some(hash);
+            } else {
+                queue.push_back((parent, hash));
+            }
+        }
+
+        None
+    }
+
+    pub fn verify(&self, root: &H256, tuples: Vec<(u32, H256)>) -> bool {
+        if let Some(h) = self.root(tuples) {
+            &h == root
+        } else {
+            false
+        }
+    }
+}
+
+pub fn build_combined_proof(tree: &[H256], mut indexes: Vec<usize>) -> Option<CombinedProof> {
+    let mut nodes = Vec::new();
+    let mut queue = VecDeque::new();
+
+    indexes.sort();
+    indexes.reverse();
+    let size = (tree.len() >> 1) + 1;
+
+    for index in indexes {
+        queue.push_back(index + size - 1);
+    }
+
+    while !queue.is_empty() {
+        let index = queue.pop_front().unwrap();
+        let sibling = calc_sibling(index);
+        if Some(&sibling) == queue.front() {
+            queue.pop_front();
+        } else if let Some(h) = tree.get(sibling) {
+            nodes.push(h.clone());
+        } else {
+            return None;
+        }
+
+        let parent = calc_parent(index);
+        if parent != 0 {
+            queue.push_back(parent);
+        }
+    }
+
+    Some(CombinedProof {
+        size: size as u32,
+        nodes,
+    })
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
These tests are difficult to determine the result by human, I prefer to use a generic `merge` function and generic type instead of hardcode `sha3` and `H256`, e.g. we can use a simple and easy to determine algorithm like integers substruction.